internal/agent/nftables_generator.go

```diff
@@ -31,19 +33,61 @@ var (
 {{ $cfg := . }}
 table {{ .FilterTableType }} {{ .FilterTableName }} {
 	chain {{ .FilterForwardChainName }} {
+		{{- range $dest := $cfg.PolicyAssignments }}
+		ct mark {{ $cfg.FWMarkBits | printf "0x%x" }} and {{ $cfg.FWMarkMask | printf "0x%x" }} ip daddr {{ $dest.Address }} goto pod-{{ $dest.Address }};
+		{{- end }}
 		ct mark {{ $cfg.FWMarkBits | printf "0x%x" }} and {{ $cfg.FWMarkMask | printf "0x%x" }} accept;
 	}
+
+	{{- range $pod := $cfg.PolicyAssignments }}
+	chain pod-{{ $pod.Address }} {
+		{{- range $pol := $pod.NetworkPolicies }}
+		jump {{ $pol }};
+		{{- end }}
+		drop;
+	}
+	{{- end }}
+
+	{{- range $policy := $cfg.NetworkPolicies }}
+	chain {{ $policy.Name }} {
+		{{- range $ruleIndex, $ingressRule := $policy.IngressRuleChains }}
+		jump {{ $policy.Name }}-rule{{ $ruleIndex }};
+		{{- end }}
+	}
+
+	{{- range $ruleIndex, $ingressRule := $policy.IngressRuleChains }}
+	chain {{ $policy.Name }}-rule{{ $ruleIndex }} {
+		{{- range $entryIndex, $entry := $ingressRule.Entries }}
+		{{ $entry.SaddrMatch.Match }} {{ $entry.PortMatch }} {{- if ne ($entry.SaddrMatch.Except | len) 0 }} jump {{ $policy.Name }}-rule{{ $ruleIndex }}-cidr{{ $entryIndex }} {{- else }} allow {{- end }};
```

Comment:
I simplified the template such that for each nftables rule (called IngressRuleChainEntry here, because "ingress rule" in the k8s sense refers to something that will result in multiple nftables rules), an address match and a port match string is generated. If there is no ipBlock, the nftables rule will apply to all address ranges, and ports respectively. An empty ingress rule will simply result in `allow`.